File: lib/jenkins/parameter.rb
require 'jenkins/pagearea'

module Jenkins
  class Parameter
    include Jenkins::PageArea

    attr_reader :name

    def self.add(job, title)
      find(:xpath, "//input[@name='parameterized']").check
      find(:xpath, "//button[text()='Add Parameter']").click
      find(:xpath, "//a[text()='#{title}']").click
      sleep 0.5
      prefix = all(:xpath, "//div[@name='parameter']").last[:path]

      return get(title).new(job, prefix)
    end

    @@types = Hash.new

    # Register Parameter type
    def self.register(title)
      @@types[title] = self
    end

    # Get type by title
    def self.get(title)
      return @@types[title]
    end

    def name=(name)
      @name = name
      control('name').set(name)
    end

    def description=(description)
      control('description').set(value)
    end

    def fill_with(value)
      raise "To be implemented by a subclass"
    end

    #Override
    def path(relative_path)
      if page.current_url.ends_with? '/configure'
        # Path on /configure page
        super relative_path
      else
        # Path on /build page
        elem = find(:xpath, "//input[@name='name' and @value='#{name}']", visible: false)
        elem[:path].gsub(/\/name$/, '') + '/' + relative_path
      end
    end
  end

  class StringParameter < Parameter

    register 'String Parameter'

    def fill_with(value)
      control('value').set value
    end

    def default=(value)
      control('defaultValue').set(value)
    end
  end
end

File: src/main/java/org/jenkinsci/test/acceptance/po/Parameter.java
package org.jenkinsci.test.acceptance.po;

import org.openqa.selenium.By;

/**
 * Parameter for builds.
 * <p>
 * Use {@link Describable} annotation to register an implementation.
 *
 * @author Kohsuke Kawaguchi
 */
public abstract class Parameter extends PageAreaImpl {
    public final Job job;
    private String name;

    public Parameter(Job job, String path) {
        super(job, path);
        this.job = job;
    }

    public String getName() {
        return name;
    }

    public Parameter setName(String name) {
        this.name = name;
        control("name").set(name);
        return this;
    }

    public Parameter setDescription(String v) {
        control("description").set(v);
        return this;
    }

    public Parameter setDefault(String value) {
        control("defaultValue").set(value);
        return this;
    }

    /**
     * Given an subtype-specific value object that represents the actual argument, fill in the parameterized build
     * form.
     */
    public abstract void fillWith(Object v);

    /**
     * Depending on whether we are on the config page or in the build page, this is different.
     */
    @Override
    public By path(String rel) {
        if (driver.getCurrentUrl().endsWith("/configure")) {
            return super.path(rel);
        }

        if (Boolean.getBoolean("new-job-page.flag.defaultValue")) {
            return by.xpath(
                    "//input[@name='name' and @value='%s']/parent::div[@name='parameter']//*[@name='%s']", name, rel);
        }

        String np = driver.findElement(by.xpath("//input[@name='name' and @value='%s']", name))
                .getAttribute("path");
        String path = np.replaceAll("/name$", "") + "/" + rel;
        return by.path(path);
    }
}
